File: app/src/main/java/io/github/fvasco/pinpoi/importer/TextImporter.java
package io.github.fvasco.pinpoi.importer;

import android.support.annotation.NonNull;
import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import io.github.fvasco.pinpoi.model.Placemark;

/**
 * Import ASCII/UTF-8 file in format
 * <ul>
 * <li>ASC: <code>longitude latitude "POI name"</code></li>
 * <li>CSV: <code>longitude, latitude, "POI name"</code></li>
 * </ul>
 *
 * @author Francesco Vasco
 */
public class TextImporter extends AbstractImporter {

    private static final Pattern LINE_PATTERN = Pattern.compile("\\s*(\"?)([+-]?\\d+\\.\\d+)\\1[,;\\s]+(\"?)([+-]?\\d+\\.\\d+)\\3[,;\\s]+\"(.*)\"\\s*");
    private static final CharsetDecoder UTF_8_DECODER = Charset.forName("UTF-8").newDecoder();

    static {
        // fails on not mappable characters
        UTF_8_DECODER.onUnmappableCharacter(CodingErrorAction.REPORT);
    }

    private final byte[] buffer = new byte[4 * 1024];

    /**
     * Decode text, if UTF-8 fails then use ISO-8859-1
     */
    public static String toString(byte[] byteBuffer, int start, int len) {
        try {
            return UTF_8_DECODER.decode(ByteBuffer.wrap(byteBuffer, start, len)).toString();
        } catch (CharacterCodingException e) {
            try {
                return new String(byteBuffer, start, len, "ISO-8859-1");
            } catch (UnsupportedEncodingException e1) {
                return new String(byteBuffer, start, len);
            }
        }
    }

    private String readLine(final InputStream inputStream) throws IOException {
        int pos = 0;
        int b;
        while ((b = inputStream.read()) >= 0) {
            if (b == '\n' || b == '\r') {
                if (pos > 0) {
                    return toString(buffer, 0, pos);
                }
            } else {
                if (pos == buffer.length) {
                    throw new IOException("Line too long");
                }
                buffer[pos] = (byte) b;
                ++pos;
            }
        }
        return b < 0 && pos == 0 ? null : toString(buffer, 0, pos);
    }


    @Override
    protected void importImpl(@NonNull final InputStream inputStream) throws IOException {
        String line;
        while ((line = readLine(inputStream)) != null) {
            final Matcher matcher = LINE_PATTERN.matcher(line);
            if (matcher.matches()) {
                try {
                    final Placemark placemark = new Placemark();
                    placemark.setLongitude(Float.parseFloat(matcher.group(2)));
                    placemark.setLatitude(Float.parseFloat(matcher.group(4)));
                    // remove double double-quotes
                    placemark.setName(matcher.group(5).replace("\"\"", "\""));
                    importPlacemark(placemark);
                } catch (final NumberFormatException nfe) {
                    Log.d(TextImporter.class.getSimpleName(), "Skip line: " + line, nfe);
                }
            } else {
                Log.d(TextImporter.class.getSimpleName(), "Skip line: " + line);
            }
        }
    }
}

File: app/src/main/java/io/github/fvasco/pinpoi/importer/TextImporter.kt
package io.github.fvasco.pinpoi.importer

import android.util.Log
import io.github.fvasco.pinpoi.model.Placemark
import io.github.fvasco.pinpoi.util.Coordinates
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.Charset
import java.nio.charset.CodingErrorAction
import java.util.regex.Pattern

/**
 * Import ASCII/UTF-8 file in format
 *
 *  * ASC: `longitude latitude "POI name"`
 *  * CSV: `longitude, latitude, "POI name"`
 *

 * @author Francesco Vasco
 */
class TextImporter : AbstractImporter() {

    private val buffer = ByteArray(4 * 1024)

    @Throws(IOException::class)
    private fun readLine(inputStream: InputStream): String? {
        var pos = 0
        var b: Int = inputStream.read()
        while (b >= 0) {
            if (b.toChar() == '\n' || b.toChar() == '\r') {
                if (pos > 0) {
                    return toString(buffer, 0, pos)
                }
            } else {
                if (pos == buffer.size) {
                    throw IOException("Line too long")
                }
                buffer[pos] = b.toByte()
                ++pos
            }
            b = inputStream.read()
        }
        return if (b < 0 && pos == 0) null else toString(buffer, 0, pos)
    }


    @Throws(IOException::class)
    override fun importImpl(inputStream: InputStream) {
        var line: String? = readLine(inputStream)
        while (line != null) {
            val matcher = LINE_PATTERN.matcher(line)
            if (matcher.matches()) {
                try {
                    val placemark = Placemark(
                            // remove double double-quotes
                            name = matcher.group(6).replace("\"\"", "\""),
                            coordinates = Coordinates(java.lang.Float.parseFloat(matcher.group(2)), java.lang.Float.parseFloat(matcher.group(4)))
                    )
                    importPlacemark(placemark)
                } catch (nfe: NumberFormatException) {
                    Log.d(TextImporter::class.java.simpleName, "Skip line: " + line, nfe)
                }

            } else {
                Log.d(TextImporter::class.java.simpleName, "Skip line: " + line)
            }
            line = readLine(inputStream)
        }
    }

    companion object {

        private val LINE_PATTERN = Pattern.compile("\\s*(\"?)([+-]?\\d+\\.\\d+)\\1[,;\\s+](\"?)([+-]?\\d+\\.\\d+)\\3[,;\\s+](\"?)(.*)\\5\\s*")
        private val UTF_8_DECODER = Charsets.UTF_8.newDecoder()
        private val LATIN1_DECODER = Charsets.ISO_8859_1.newDecoder()

        init {
            // fails on not mappable characters
            UTF_8_DECODER.onUnmappableCharacter(CodingErrorAction.REPORT)
        }

        /**
         * Decode text, if UTF-8 fails then use ISO-8859-1
         */
        fun toString(byteBuffer: ByteArray, start: Int, len: Int): String =
            try {
                UTF_8_DECODER.decode(ByteBuffer.wrap(byteBuffer, start, len)).toString()
            } catch (e: CharacterCodingException) {
                LATIN1_DECODER.decode(ByteBuffer.wrap(byteBuffer, start, len)).toString()
            }
    }
}
